src/compiler/tsc.ts

```diff
@@ -146,131 +173,176 @@ module ts {
-     * Upon detecting a file change, watchProgram will queue up file modification events for the next
-     * 250ms and then perform a recompilation. The reasoning is that in some cases, an editor can
-     * save all files at once, and we'd like to just perform a single recompilation.
-     */
-    function watchProgram(commandLine: ParsedCommandLine, compilerHost: CompilerHost): void {
-        var watchers: Map<FileWatcher> = {};
-        var updatedFiles: Map<boolean> = {};
-
-        // Compile the program the first time and watch all given/referenced files.
-        var program = compile(commandLine, compilerHost).program;
-        reportDiagnostic(createCompilerDiagnostic(Diagnostics.Compilation_complete_Watching_for_file_changes));
-        addWatchers(program);
-        return;
-
-        function addWatchers(program: Program) {
-            forEach(program.getSourceFiles(), f => {
-                var filename = getCanonicalName(f.filename);
-                watchers[filename] = sys.watchFile(filename, fileUpdated);
-            });
+            if (configFilename) {
+                configFileWatcher = sys.watchFile(configFilename, configFileChanged);
+            }
         }
 
-        function removeWatchers(program: Program) {
-            forEach(program.getSourceFiles(), f => {
-                var filename = getCanonicalName(f.filename);
-                if (hasProperty(watchers, filename)) {
-                    watchers[filename].close();
+        performCompilation();
+
+        // Invoked to perform initial compilation or re-compilation in watch mode
+        function performCompilation() {
+
+            if (!cachedSourceFiles) {
+                if (configFilename) {
+                    var configObject = readConfigFile(configFilename);
+                    if (!configObject) {
+                        reportDiagnostic(createCompilerDiagnostic(Diagnostics.Unable_to_open_file_0, configFilename));
+                        return sys.exit(EmitReturnStatus.CompilerOptionsErrors);
+                    }
+                    var configParseResult = parseConfigFile(configObject, getDirectoryPath(configFilename));
+                    if (configParseResult.errors.length > 0) {
+                        reportDiagnostics(configParseResult.errors);
+                        return sys.exit(EmitReturnStatus.CompilerOptionsErrors);
+                    }
+                    rootFilenames = configParseResult.filenames;
+                    compilerOptions = extend(commandLine.options, configParseResult.options);
                 }
-            });
+                else {
+                    rootFilenames = commandLine.filenames;
+                    compilerOptions = commandLine.options;
+                }
+                compilerHost = createCompilerHost(compilerOptions);
+                hostGetSourceFile = compilerHost.getSourceFile;
+                compilerHost.getSourceFile = getSourceFile;

```

Comment:
What I meant was, it's not clear what is the difference between hostGetSourceFile and getSourceFile. Can you rename them so that each sounds like it excludes the other?